rclcpp/test/test_parameter.cpp

```diff
@@ -0,0 +1,551 @@
+
+  rclcpp::parameter::ParameterVariant long_variant("integer_param", TEST_VALUE_L);
+  EXPECT_EQ(long_variant.get_name(), "integer_param");
+  EXPECT_EQ(long_variant.get_type(), rclcpp::parameter::ParameterType::PARAMETER_INTEGER);
+  EXPECT_EQ(long_variant.get_type_name(), "integer");
+  EXPECT_EQ(long_variant.get_value<rclcpp::parameter::ParameterType::PARAMETER_INTEGER>(),
+    TEST_VALUE_L);
+  EXPECT_EQ(long_variant.get_parameter_value().integer_value, TEST_VALUE_L);
+  EXPECT_EQ(long_variant.get_parameter_value().type,
+    rcl_interfaces::msg::ParameterType::PARAMETER_INTEGER);
+
+  rcl_interfaces::msg::Parameter integer_param = long_variant.to_parameter();
+  EXPECT_EQ(integer_param.name, "integer_param");
+  EXPECT_EQ(integer_param.value.type, rcl_interfaces::msg::ParameterType::PARAMETER_INTEGER);
+  EXPECT_EQ(integer_param.value.integer_value, TEST_VALUE_L);
+
+  // From parameter message
+  rclcpp::parameter::ParameterVariant from_msg =
+    rclcpp::parameter::ParameterVariant::from_parameter(integer_param);
+  EXPECT_EQ(from_msg.get_name(), "integer_param");
+  EXPECT_EQ(from_msg.get_type(), rclcpp::parameter::ParameterType::PARAMETER_INTEGER);
+  EXPECT_EQ(from_msg.get_type_name(), "integer");
+  EXPECT_EQ(from_msg.get_value<rclcpp::parameter::ParameterType::PARAMETER_INTEGER>(),
+    TEST_VALUE_L);
+  EXPECT_EQ(from_msg.get_parameter_value().integer_value, TEST_VALUE_L);
+  EXPECT_EQ(from_msg.get_parameter_value().type,
+    rcl_interfaces::msg::ParameterType::PARAMETER_INTEGER);
+}
+
+TEST(TestParameter, double_variant) {
+  const double TEST_VALUE {-42.1};
+
+  // Direct instantiation
+  rclcpp::parameter::ParameterVariant double_variant("double_param", TEST_VALUE);
+  EXPECT_EQ(double_variant.get_name(), "double_param");
+  EXPECT_EQ(double_variant.get_type(), rclcpp::parameter::ParameterType::PARAMETER_DOUBLE);
+  EXPECT_EQ(double_variant.get_type_name(), "double");
+  EXPECT_EQ(double_variant.get_value<rclcpp::parameter::ParameterType::PARAMETER_DOUBLE>(),
+    TEST_VALUE);
+  EXPECT_EQ(double_variant.get_parameter_value().double_value, TEST_VALUE);
+  EXPECT_EQ(double_variant.get_parameter_value().type,
+    rcl_interfaces::msg::ParameterType::PARAMETER_DOUBLE);
+  EXPECT_EQ(double_variant.as_double(), TEST_VALUE);
+
+  EXPECT_THROW(double_variant.as_bool(), std::runtime_error);
+  EXPECT_THROW(double_variant.as_int(), std::runtime_error);
+  EXPECT_THROW(double_variant.as_string(), std::runtime_error);
+  EXPECT_THROW(double_variant.as_byte_array(), std::runtime_error);
+  EXPECT_THROW(double_variant.as_bool_array(), std::runtime_error);
+  EXPECT_THROW(double_variant.as_integer_array(), std::runtime_error);
+  EXPECT_THROW(double_variant.as_double_array(), std::runtime_error);
+  EXPECT_THROW(double_variant.as_string_array(), std::runtime_error);
+
+  EXPECT_EQ(double_variant.value_to_string(), "-42.100000");
+
+  const float TEST_VALUE_F {static_cast<float>(-TEST_VALUE)};
```

Comment:
This is in here because float and double arrays get mapped to the same parameter type. I meant to follow the same pattern as above, with ints and long ints. Will review.

I was also trying to kill two birds with one stone: check that floats work, but also check for negative values.

Would you prefer that all variants, even when mapped to the same array type, have their own tests?